src/NuGet.Clients/NuGet.SolutionRestoreManager/BaseInfoBar.cs

```diff
@@ -0,0 +1,163 @@
+            {
+                return null;
+            }
+
+            return tempObject as IVsInfoBarHost;
+        }
+
+        internal async Task ShowAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (_visible)
+            {
+                return;
+            }
+
+            try
+            {
+                var infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    return;
+                }
+
+                var infoBarHost = await GetInfoBarHostAsync(cancellationToken);
+                if (infoBarHost == null)
+                {
+                    return;
+                }
+
+                // Ensure that we are on the UI thread before interacting with the UI
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+                var infoBarModel = GetInfoBarModel();
+
+                _infoBarUIElement = infoBarFactory.CreateInfoBar(infoBarModel);
+                _infoBarUIElement.Advise(this, out uint cookie);
+                _eventCookie = cookie;
+
+                infoBarHost.AddInfoBar(_infoBarUIElement);
+
+                _visible = true;
+            }
+            catch
+            {
+                return;
+            }
+        }
+
+        internal async Task HideAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                // Ensure that we are on the UI thread before interacting with the UI element
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+                if (!_visible)
+                {
+                    return;
+                }
+
+                _closeFromHide = true;
+
+                try
+                {
+                    _infoBarUIElement?.Close();
+                    _infoBarUIElement = null;
+                }
+                finally
+                {
+                    _closeFromHide = false;
+                }
+
+            }
+            catch
+            {
+                return;
+            }
+        }
+
+        public void OnClosed(IVsInfoBarUIElement infoBarUIElement)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_eventCookie.HasValue)
+            {
+                infoBarUIElement?.Unadvise(_eventCookie.Value);
+                _eventCookie = null;
+            }
+
+            _visible = false;
+            OnClosed(_closeFromHide);
+        }
+
+        public void OnActionItemClicked(IVsInfoBarUIElement infoBarUIElement, IVsInfoBarActionItem actionItem)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            try
+            {
+                if (actionItem.ActionContext is string action)
```

Comment:
Instead of creating and loading the PM UI manually, I think we should try having the action invoke an existing command as if the context menu was used (eg, solution explorer or Tools-options)

Reply:
I'd rather we try to pass an argument through the action before duplicating UI creation logic. If it's absolutely not possible to differentiate telemetry any other way, then we can consider that strategy so we can detect internal vs. external.

Reply:
An idea how you can avoid an RPS regression and loading of extra dlls unnecessarily you can do some declaration vs implementation trickery. 

You can declare a service interface somewhere in NuGet.VisualStudio.Common that can be used to start the PM UI. 
Then you declare that implementation in NuGet.Tools and only lazily initialize that service in NuGet.SolutionRestoreManager so that it's only initialized  when someone clicks on it. In the end we load the NuGet.Tools dll when someone clicks on it.